torch/csrc/distributed/rpc/ProcessGroupAgent.cpp

```diff
@@ -143,48 +137,31 @@ std::shared_ptr<FutureMessage> ProcessGroupAgent::send(
 }
 
 void ProcessGroupAgent::enqueue(SendWork work) {
-  std::unique_lock<std::mutex> lock(sendQueueMutex_);
-  sendQueue_.emplace_back(std::move(work));
-  lock.unlock();
-
-  workProduceCV_.notify_one();
-}
-
-// making sure tensors are not deleted before send finishes
-void ProcessGroupAgent::sendLoop() {
-  std::unique_lock<std::mutex> lock(sendQueueMutex_);
-
-  while (!stop_) {
-    if (sendQueue_.empty()) {
-      workProduceCV_.wait(lock);
-      continue;
-    }
-
-    auto work = std::move(sendQueue_.front());
-    sendQueue_.pop_front();
-    lock.unlock();
-
-    workConsumeCV_.notify_one();
-
-
-    std::stringstream ss;
-    serialize(work.message_, ss);
-    std::string str = ss.str();
-
-    std::vector<torch::Tensor> preamble = {
-      torch::tensor(
-        {
-          (int64_t)pg_->getRank(),
-          (int64_t)str.length(),
-        }, {torch::kLong})
-    };
-    pg_->send(preamble, work.dstRank_, work.dstRank_ /* channelTag */)->wait();
-    std::vector<torch::Tensor> payload =
-        {torch::from_blob((void *)str.c_str(), str.length(), {torch::kChar})};
-    pg_->send(payload, work.dstRank_, work.dstRank_ /* channelTag */)->wait();
-
-    lock.lock();
-  }
+  sendThreadPool_.run(std::bind(
```

Comment:
Hey @ilia-cher, it seems `ThreadPool::run` made a [copy](https://github.com/pytorch/pytorch/blob/82aecfad6a0a1d7790f7e96157033bcc576d346e/c10/core/thread_pool.h#L57) of the `std::function`. Is there any way to avoid this copy? Copy the bound work here can be expensive.

Reply:
it does make a copy to store the task internally, I'm not sure what better approach you'd suggest here, in any way we should demonstrate the issue and the win using a micro-benchmark